feat: validate server options

olevski opened the pull request:
closes #71 

So I implemented evaluation on the helm chart values but only for the server options portion. I will add a new issue to add it for the whole values.yaml file.

I also added more detailed validation (in line with the values.yaml validation) for the API.

I did some digging online to see if marshmallow can parse jsonschema and convert to marshmallow type schemas and it cannot unfortunately. There is a library that can convert a marshmallow schema to jsonschema but it cannot include the custom validation we need for some of the server options stuff. So I think we unfortunately have to have this kind of duplication.

This is deployed at https://tasko.dev.renku.ch/

olevski commented:
issue for adding validation for the whole values.yaml file #530

ableuler reviewed (changes requested):
Thanks for the careful work on validating that the server options conform to the desired schema. I agree that the duplication is a bit annoying, but we can live with it. 

The really important point still missing here (you might consider this simply a bug of the initial implementation of those server options) is an actual check if the server options which come as part as a POST body are elements of the options that were offered for selection in the first place. Basically, we have to prevent this

![image](https://user-images.githubusercontent.com/13014091/106636632-659eec80-6582-11eb-8b71-6a8886c5b51c.png)

when the dialog only offers this

![image](https://user-images.githubusercontent.com/13014091/106636974-90894080-6582-11eb-9d4a-76cc80b3083c.png).

I simply replayed the POST request from the browser while giving myself a bit more CPU for this.

olevski commented on renku_notebooks/api/custom_fields.py:
```diff
@@ -48,3 +49,13 @@ def _deserialize(
         raise ValidationError(errors)
+
+
+serverOptionCpuValue = fields.Number(
```
So the problem with these validations is as follows. Kubernetes reports the following as the regex used to validate them: 
`^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$`. However this lets 'through' some additional values that are then further checked after they pass the initial regex check. For example according to this regex a value of `-100M` is a valid for memory and a few other weird cases but these will not pass and a different error message will be shown.

To simplify things I decided to:
- just use regex for validation
- using regex only for validation allows me to easily encode this in the values.yaml validation and marshmallow, I can easily write fancy stuff in python for marshmallow but that is not the case for the helm values.yaml validation so I want to keep things simple
- with this regex capture the subset of valid options that look good in the UI and are easy to in the json-schema
- these 'look-good' valid options are then:
  - cpu: number greater than zero with at most three decimals (that is the the most precision k8s does for cpu)
  - memory: whole number followed by either E,P,T,G,M,K, or Ei,Pi,Ti,Gi,Mi,Ki
  - gpu: whole number equal to zero or greater
 
I think this still covers most cases. What do you think?

ableuler replied:
Ok, makes sense 👍

olevski commented on renku_notebooks/api/custom_fields.py:
```diff
@@ -48,3 +49,13 @@ def _deserialize(
+    validate=lambda x: x > 0.0 and (x % 1 >= 0.001 or x % 1 == 0.0), required=True,
+)
+serverOptionMemoryValue = fields.String(
+    validate=lambda x: re.match(r"^(?:[1-9][0-9]*|[0-9]\.[0-9]*)[EPTGMK][i]{0,1}$", x)
```
k8s understands both. G is a power of 10 consistent with SI prefixes in general, while Gi is a power of 2. So for example 1K = 1000 bytes, 1Ki = 1024 bytes.

ableuler replied:
oh true, forgot about this distinction! I think I'll never get used to the fact that computer scientists thought that it was convenient to denote 1024 simply as 1000...

olevski commented:
@ableuler let me know what you think about my responses. Also I added validation for the server options in the post request for starting a server as you suggested. I am sorry I missed that. I also added a test to ensure that trying to start a server with bad options will fail.

ableuler commented on renku_notebooks/api/schemas.py:
```diff
@@ -1,8 +1,45 @@
+    def validate_server_options(self, data, **kwargs):
+        server_options = read_server_options_file()
+        for option in data.keys():
+            if server_options.get(option, {}).get("type", None) == "boolean":
```
This does the right thing, but it took me a moment to process the logic. How about something like this?
``
if option not in server_options:
  continue
if server_options[option]["type"] == "boolean":
  continue
if data[option] not in server_options[option]["options"]:
  raise ValidationError...
 ```

olevski replied:
Ok sure I will keep the `get` statements to avoid errors when a key is missing.

ableuler replied:
Don't server options always have a type and non-boolean server options always have options? Wouldn't in this case uncaught KeyError be more helpful in detecting a problem here than silently using an empty dictionary or list?

olevski replied:
I typed too soon. I think I can fully get away from the `get` statements with what you propose.

olevski replied:
@ableuler you are right. The `get` statements are gone.

olevski commented:
Ok great thanks. Also ran the acceptance tests locally with docker and they passed.
